### PR TITLE
Write macro named arguments

### PR DESCRIPTION
**Pull request by botika**

### CODE REVIEW

**Comment by djc** [comment]
What's the perf win here?

**Comment by botika** [comment]
```
 Running target/release/deps/all-1cd74d22a4e3eac1
Big table               time:   [709.48 us 710.21 us 711.05 us]
                        change: [-1.2486% -0.5887% +0.0479%] (p = 0.08 > 0.05)
                        No change in performance detected.
Found 13 outliers among 100 measurements (13.00%)
  5 (5.00%) high mild
  8 (8.00%) high severe

Teams                   time:   [847.99 ns 849.00 ns 850.18 ns]
                        change: [-2.1367% -1.2890% -0.5289%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 14 outliers among 100 measurements (14.00%)
  5 (5.00%) high mild
  9 (9.00%) high severe
```

**Comment by botika** [comment]
It is minimal because this case is very little reflected in the bench.

File: askama_derive/src/generator.rs [hunk 1]
```diff
@@ -638,18 +642,30 @@ impl<'a> Generator<'a> {
...
-
-                            buf_format.write("{}");
-                            buf_expr.write("&");
-                            buf_expr.write(&match (wrapped, &self.input.escaping) {
-                                (Wrapped, &Html) | (Wrapped, &None) | (Unwrapped, &None) => {
-                                    expr_buf.buf
-                                }
-                                (Unwrapped, &Html) => {
-                                    format!("::askama::MarkupDisplay::from(&{})", expr_buf.buf)
+                            if cache_id.contains_key(&expr_buf) {
```
Comment (djc) [inline askama_derive/src/generator.rs]: Let's match against `cache_id.entry()` here, which should be more efficient (single cache lookup if it is present).

File: askama_derive/src/generator.rs [hunk 2]
```diff
@@ -628,6 +631,7 @@ impl<'a> Generator<'a> {
             } else {
                 let mut buf_format = Buffer::new(0);
                 let mut buf_expr = Buffer::new(buf.indent + 1);
+                let mut cache_id = HashMap::with_capacity(self.buf_writable.len());
```
Comment (djc) [inline askama_derive/src/generator.rs]: Let's call this `expr_cache`?

File: askama_derive/src/generator.rs [hunk 3]
```diff
@@ -638,18 +642,30 @@ impl<'a> Generator<'a> {
...
+                                            expr_buf.buf
+                                        ));
+                                    }
                                 }
-                            });
-                            buf_expr.writeln(",");
+
+                                buf_expr.writeln(",");
+                                buf_format.write(&format!("{{a{}}}", id));
+                                cache_id.insert(expr_buf.clone(), id);
```
Comment (djc) [inline askama_derive/src/generator.rs]: Move the `insert()` and `named += 1` up to right after the line `let id = self.named` so the logic is better localized?

Actually, we can probably move ownership of `expr_buf` into the cache map? In that case, it needs to be down at the bottom, but we should get rid of the `clone()` (and the corresponding `derive()`).

File: askama_derive/src/generator.rs (hunk 2 above)
Comment (botika) [reply askama_derive/src/generator.rs]: Ok

File: askama_derive/src/generator.rs (hunk 1 above)
Comment (botika) [reply askama_derive/src/generator.rs]: Ok

File: askama_derive/src/generator.rs (hunk 3 above)
Comment (botika) [reply askama_derive/src/generator.rs]: and ok

**Pull request merged**